File: engineering/engineering/page/tyre_dashboard/tyre_dashboard.py
from calendar import monthrange
from collections import defaultdict
from datetime import date, timedelta
import json
from pathlib import Path

import frappe
from frappe.utils import add_months, getdate, today

from engineering.tyre_analytics import get_latest_analytics


STANDARD_LAYOUT_IMAGE = (
    "/assets/engineering/images/tyre_layouts/Standard_adt.png?v=20260716b"
)
B60E_LAYOUT_IMAGE = (
    "/assets/engineering/images/tyre_layouts/Bell_B60E.png?v=20260716b"
)
B60E_POSITIONS = {"RRO", "RRI", "LRI", "LRO"}
MOCK_DAMAGE_MARKER = "[MOCK DAMAGE: GRADER-COMPARISON-2026]"
MOCK_GRADER_DATA_FILE = (
    Path(__file__).resolve().parents[3]
    / "mock_grader_utilisation.json"
)
MODEL_FIELDS = (
    "custom_vehicle_model",
    "vehicle_model",
    "custom_model",
    "model",
)
MAKE_FIELDS = (
    "custom_vehicle_make",
    "vehicle_make",
    "custom_make",
    "make",
    "manufacturer",
    "brand",
)
KNOWN_MODELS = ("740GC", "B60E", "B60", "B45E", "B40E", "B40D", "A40G")


def _month_start(value):
    value = getdate(value)
    return date(value.year, value.month, 1)


def _month_end(value):
    return date(value.year, value.month, monthrange(value.year, value.month)[1])


def _average(values, precision=1):
    values = [value for value in values if value is not None]
    return round(sum(values) / len(values), precision) if values else 0


def _first_doc_value(doc, fieldnames):
    meta = frappe.get_meta(doc.doctype)

    for fieldname in fieldnames:
        if meta.has_field(fieldname) and doc.get(fieldname):
            return str(doc.get(fieldname)).strip()

    return ""


def _asset_information(asset_name):
    if not frappe.db.exists("Asset", asset_name):
        return {"vehicle_make": "", "model": ""}

    asset = frappe.get_doc("Asset", asset_name)
    item = None
    vehicle_make = _first_doc_value(asset, MAKE_FIELDS)
    model = _first_doc_value(asset, MODEL_FIELDS)
    searchable = [
        asset.name,
        asset.get("item_name"),
        asset.get("item_code"),
        vehicle_make,
        model,
    ]

    if asset.get("item_code") and frappe.db.exists("Item", asset.item_code):
        item = frappe.get_doc("Item", asset.item_code)
        vehicle_make = vehicle_make or _first_doc_value(item, MAKE_FIELDS)
        model = model or _first_doc_value(item, MODEL_FIELDS)
        searchable.extend(
            [
                item.get("item_name"),
                item.get("item_code"),
                item.get("description"),
                vehicle_make,
                model,
            ]
        )

    model_text = "".join(
        character
        for character in " ".join(
            str(value) for value in searchable if value not in (None, "")
        ).upper()
        if character.isalnum()
    )

    for known_model in KNOWN_MODELS:
        if known_model in model_text:
            model = known_model
            break

    if not vehicle_make and str(model).upper().startswith("B"):
        vehicle_make = "Bell"

    return {"vehicle_make": vehicle_make, "model": model}


def _group_average(rows, key_field, value_field):
    grouped = defaultdict(list)

    for row in rows:
        value = row.get(value_field)

        if value is not None and value > 0:
            grouped[row.get(key_field) or "Not Set"].append(value)

    result = [
        {
            "label": label,
            "value": _average(values, 2),
        }
        for label, values in grouped.items()
    ]
    result.sort(key=lambda item: (-item["value"], item["label"]))
    return result


def _pressure_compliance_by_site(rows):
    grouped = defaultdict(list)

    for row in rows:
        grouped[row.site or "Not Set"].append(row.pressure_compliant)

    result = [
        {
            "label": site,
            "value": round((sum(values) / len(values)) * 100, 1),
        }
        for site, values in grouped.items()
    ]
    result.sort(key=lambda item: item["label"])
    return result


def _replacement_forecast(rows, anchor_date):
    first_month = _month_start(add_months(anchor_date, 1))
    forecast = []

    for offset in range(5):
        month = _month_start(add_months(first_month, offset))
        last_day = _month_end(month)
        count = sum(
            1
            for row in rows
            if row.replacement_date
            and month <= getdate(row.replacement_date) <= last_day
        )
        forecast.append(
            {
                "label": month.strftime("%b %Y"),
                "value": count,
            }
        )

    return forecast


def _mock_grader_damage_comparison():
    """Return only the isolated mock comparison dataset."""

    if not MOCK_GRADER_DATA_FILE.exists():
        return []

    with MOCK_GRADER_DATA_FILE.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)

    survey_meta = frappe.get_meta("Tyre Survey")
    tyres_field = survey_meta.get_field("tyres")

    if not tyres_field or not tyres_field.options:
        return []

    child_table = "tab{0}".format(
        tyres_field.options.replace("`", "``")
    )
    damage_rows = frappe.db.sql(
        """
        SELECT ts.survey_date
        FROM `{child_table}` tsi
        INNER JOIN `tabTyre Survey` ts ON ts.name = tsi.parent
        WHERE ts.inspector_name = %(inspector)s
          AND ts.docstatus = 0
          AND COALESCE(tsi.condition_notes, '') LIKE %(marker)s
        """.format(child_table=child_table),
        {
            "inspector": "Mock Survey Generator",
            "marker": "%{0}%".format(MOCK_DAMAGE_MARKER),
        },
        as_dict=True,
    )
    damage_counts = defaultdict(int)

    for row in damage_rows:
        month_key = getdate(row.survey_date).strftime("%Y-%m")
        damage_counts[month_key] += 1

    return [
        {
            "label": date.fromisoformat(
                "{0}-01".format(month["month"])
            ).strftime("%b %Y"),
            "month": month["month"],
            "grader_utilisation": month.get("grader_utilisation") or 0,
            "damaged_tyres": damage_counts.get(month["month"], 0),
            "damage_target": month.get("damage_target") or 0,
            "data_status": "MOCK ONLY",
        }
        for month in payload.get("months") or []
    ]


def _urgent_rows(rows):
    sorted_rows = sorted(
        rows,
        key=lambda row: (
            -row.urgency_score,
            row.remaining_months
            if row.remaining_months is not None
            else 999999,
            row.site or "",
            row.fleet_number or "",
        ),
    )

    return [
        {
            "site": row.site,
            "fleet_number": row.fleet_number,
            "position": row.position,
            "serial_number": row.serial_number,
            "average_rtd": row.average_rtd,
            "wear_rate": row.wear_rate,
            "remaining_months": row.remaining_months,
            "replacement_date": row.replacement_date,
            "urgency_score": row.urgency_score,
            "urgency_band": row.urgency_band,
            "required_action": row.required_action,
        }
        for row in sorted_rows[:10]
    ]


def _adt_view(rows, fleet_number):
    if not fleet_number:
        return {
            "fleet_number": None,
            "site": None,
            "vehicle_make": None,
            "model": None,
            "tyre_layout": "standard",
            "layout_image": STANDARD_LAYOUT_IMAGE,
            "tyres": [],
        }

    tyres = [row for row in rows if row.fleet_number == fleet_number]
    positions = {str(row.position or "").upper() for row in tyres}

    asset_information = _asset_information(fleet_number)

    model_code = "".join(
        character
        for character in str(asset_information.get("model") or "").upper()
        if character.isalnum()
    )
    is_b60 = "B60" in model_code or bool(
        positions.intersection(B60E_POSITIONS)
    )
    layout_image = (
        B60E_LAYOUT_IMAGE
        if is_b60
        else STANDARD_LAYOUT_IMAGE
    )

    return {
        "fleet_number": fleet_number,
        "site": tyres[0].site if tyres else None,
        "vehicle_make": asset_information.get("vehicle_make"),
        "model": asset_information.get("model"),
        "tyre_layout": "b60e" if is_b60 else "standard",
        "layout_image": layout_image,
        "tyres": [
            {
                "survey": row.survey,
                "survey_date": row.survey_date,
                "position": row.position,
                "serial_number": row.serial_number,
                "tyre_make": row.tyre_make,
                "tread_pattern": row.tread_pattern,
                "otd": row.otd,
                "average_rtd": row.average_rtd,
                "rtd_percent": row.rtd_percent,
                "wear_rate": row.wear_rate,
                "recommended_pressure": row.recommended_pressure,
                "actual_pressure": row.actual_pressure,
                "pressure_variance": row.pressure_variance,
                "scrap_limit_mm": 14,
                "remaining_months": row.remaining_months,
                "replacement_date": row.replacement_date,
                "urgency_score": row.urgency_score,
                "urgency_band": row.urgency_band,
                "condition_notes": row.condition_notes,
                "required_action": row.required_action,
            }
            for row in tyres
        ],
    }


@frappe.whitelist()
def get_dashboard_data(as_on_date=None, site=None, fleet_number=None):
    anchor_date = getdate(as_on_date or today())
    all_rows = get_latest_analytics(
        as_on_date=anchor_date,
        include_mock=True,
    )
    rows = all_rows

    if site:
        rows = [row for row in rows if row.site == site]

    bands = ["Critical", "Warning", "Check", "Good", "Normal"]
    band_counts = [
        {
            "label": band,
            "value": sum(1 for row in rows if row.urgency_band == band),
        }
        for band in bands
    ]
    due_date = anchor_date + timedelta(days=30)

    return {
        "as_on_date": anchor_date,
        "scrap_limit_mm": 14,
        "kpis": {
            "tyres_analysed": len(rows),
            "critical": sum(1 for row in rows if row.urgency_band == "Critical"),
            "warning": sum(1 for row in rows if row.urgency_band == "Warning"),
            "below_25_percent": sum(1 for row in rows if row.rtd_percent < 25),
            "underinflated": sum(1 for row in rows if row.pressure_variance < -10),
            "due_within_30_days": sum(
                1
                for row in rows
                if row.replacement_date
                and getdate(row.replacement_date) <= due_date
            ),
        },
        "replacement_forecast": _replacement_forecast(rows, anchor_date),
        "urgency_bands": band_counts,
        "site_wear": _group_average(rows, "site", "wear_rate"),
        "brand_wear": _group_average(rows, "tyre_make", "wear_rate")[:10],
        "pressure_compliance": _pressure_compliance_by_site(rows),
        "mock_grader_damage": _mock_grader_damage_comparison(),
        "urgent_tyres": _urgent_rows(rows),
        # The machine view is intentionally independent of the Site filter.
        # A selected ADT must always show its latest fitted tyres and layout.
        "adt_view": _adt_view(all_rows, fleet_number),
    }

File: engineering/engineering/page/tyre_dashboard/tyre_dashboard.js
frappe.pages["tyre-dashboard"].on_page_load = function (wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __("Tyre Management Dashboard"),
        single_column: true,
    });

    const asOnDate = page.add_field({
        fieldname: "as_on_date",
        label: __("As On Date"),
        fieldtype: "Date",
        default: frappe.datetime.get_today(),
    });
    const site = page.add_field({
        fieldname: "site",
        label: __("Site"),
        fieldtype: "Link",
        options: "Location",
    });
    const fleetNumber = page.add_field({
        fieldname: "fleet_number",
        label: __("ADT Six-Wheel View"),
        fieldtype: "Link",
        options: "Asset",
        get_query() {
            return { filters: { asset_category: "ADT" } };
        },
    });

    page.set_primary_action(__("Refresh"), loadDashboard, "refresh");
    page.add_inner_button(__("Urgency Report"), () => {
        frappe.set_route("query-report", "Tyre Urgency Report");
    });
    page.add_inner_button(__("Replacement Forecast"), () => {
        frappe.set_route("query-report", "Tyre Replacement Forecast");
    });
    page.add_inner_button(__("ADT Detail Report"), () => {
        frappe.set_route("query-report", "ADT Six Wheel Status");
    });

    const $container = $("<div class='tyre-dashboard'></div>").appendTo(page.main);

    $container.html(`
        <style>
            .tyre-dashboard { padding: 18px 4px 40px; }
            .tyre-dashboard .dashboard-note { color: var(--text-muted); margin-bottom: 14px; }
            .tyre-kpis { display: grid; grid-template-columns: repeat(6, minmax(130px, 1fr)); gap: 12px; margin-bottom: 18px; }
            .tyre-kpi { background: var(--card-bg); border: 1px solid var(--border-color); border-radius: 10px; padding: 14px; min-height: 92px; }
            .tyre-kpi .value { font-size: 28px; font-weight: 700; line-height: 1.1; }
            .tyre-kpi .label { color: var(--text-muted); margin-top: 8px; font-size: 12px; }
            .tyre-kpi.red { border-top: 4px solid #d32f2f; }
            .tyre-kpi.orange { border-top: 4px solid #f57c00; }
            .tyre-kpi.blue { border-top: 4px solid #1565c0; }
            .tyre-kpi.green { border-top: 4px solid #2e7d32; }
            .tyre-chart-grid { display: grid; grid-template-columns: repeat(2, minmax(360px, 1fr)); gap: 14px; margin-bottom: 18px; }
            .tyre-panel { background: var(--card-bg); border: 1px solid var(--border-color); border-radius: 10px; padding: 16px; }
            .tyre-panel.wide { grid-column: 1 / -1; }
            .tyre-panel h4 { margin: 0 0 12px; font-size: 15px; font-weight: 600; }
            .tyre-chart { min-height: 270px; }
            .adt-view { margin-bottom: 18px; }
            .adt-layout { display: grid; grid-template-columns: minmax(245px, 1fr) minmax(240px, 0.8fr) minmax(245px, 1fr); gap: 18px; align-items: center; }
            .adt-layout-image { width: 100%; max-height: 620px; object-fit: contain; }
            .adt-card-column { display: flex; flex-direction: column; gap: 10px; }
            .adt-tyre-card { border: 1px solid var(--border-color); border-left: 5px solid #1565c0; border-radius: 8px; padding: 11px; background: var(--fg-color); }
            .adt-tyre-card.critical { border-left-color: #d32f2f; }
            .adt-tyre-card.warning { border-left-color: #f57c00; }
            .adt-tyre-card.check { border-left-color: #fbc02d; }
            .adt-tyre-card.good { border-left-color: #2e7d32; }
            .adt-tyre-card .title { font-weight: 700; margin-bottom: 5px; }
            .adt-tyre-card .spec-list { color: var(--text-muted); font-size: 12px; line-height: 1.55; margin: 7px 0 0; padding-left: 19px; }
            .adt-tyre-card .spec-list li { margin: 1px 0; }
            .adt-tyre-card .spec-list strong { color: var(--text-color); font-weight: 600; }
            .urgent-table { width: 100%; border-collapse: collapse; }
            .urgent-table th, .urgent-table td { border-bottom: 1px solid var(--border-color); padding: 8px; text-align: left; font-size: 12px; }
            .urgent-table th { color: var(--text-muted); font-weight: 600; }
            .band { font-weight: 700; }
            .band.critical { color: #d32f2f; }
            .band.warning { color: #f57c00; }
            .band.check { color: #b8860b; }
            .dashboard-empty { padding: 30px; text-align: center; color: var(--text-muted); }
            @media (max-width: 1100px) {
                .tyre-kpis { grid-template-columns: repeat(3, 1fr); }
                .adt-layout { grid-template-columns: 1fr; }
                .adt-layout-image { max-height: 300px; order: -1; }
            }
            @media (max-width: 760px) {
                .tyre-kpis, .tyre-chart-grid { grid-template-columns: 1fr; }
            }
        </style>
        <div class="dashboard-note">${__("Latest survey condition and forecast using a 14 mm scrap limit.")}</div>
        <div class="tyre-kpis" id="tyre-kpis"></div>
        <div class="tyre-panel adt-view">
            <h4>${__("ADT Six-Wheel Condition")}</h4>
            <div id="adt-layout"></div>
        </div>
        <div class="tyre-chart-grid">
            <div class="tyre-panel"><h4>${__("Five-Month Replacement Forecast")}</h4><div id="forecast-chart" class="tyre-chart"></div></div>
            <div class="tyre-panel"><h4>${__("Urgency Distribution")}</h4><div id="urgency-chart" class="tyre-chart"></div></div>
            <div class="tyre-panel"><h4>${__("Average Wear by Site")}</h4><div id="site-wear-chart" class="tyre-chart"></div></div>
            <div class="tyre-panel"><h4>${__("Average Wear by Brand")}</h4><div id="brand-wear-chart" class="tyre-chart"></div></div>
            <div class="tyre-panel"><h4>${__("Pressure Compliance by Site")}</h4><div id="pressure-chart" class="tyre-chart"></div></div>
            <div class="tyre-panel wide">
                <h4>${__("Mock Grader Utilisation vs Mock Tyre Damage")}</h4>
                <div class="dashboard-note">${__("Development-only synthetic comparison. This panel does not use or alter actual Availability and Utilisation records.")}</div>
                <div id="mock-grader-damage-chart" class="tyre-chart"></div>
            </div>
        </div>
        <div class="tyre-panel">
            <h4>${__("Top 10 Urgent Tyres")}</h4>
            <div id="urgent-list"></div>
        </div>
    `);

    asOnDate.$input.on("change", loadDashboard);
    site.$input.on("change", loadDashboard);
    fleetNumber.$input.attr(
        "placeholder",
        __("Type or select an ADT fleet number, then press Enter")
    );
    fleetNumber.$input.on("keyup.tyre-dashboard-enter", event => {
        if (event.key !== "Enter" && event.keyCode !== 13) {
            return;
        }

        // Enter selects the highlighted Link result on keydown. Waiting until
        // keyup lets Frappe commit and validate the Asset before we read it.
        window.setTimeout(loadDashboard, 0);
    });

    function escape(value) {
        return frappe.utils.escape_html(String(value ?? ""));
    }

    function displayDate(value) {
        return value ? frappe.datetime.str_to_user(value) : "—";
    }

    function bandClass(value) {
        return String(value || "normal").toLowerCase();
    }

    function renderKpis(kpis) {
        const cards = [
            [kpis.tyres_analysed, __("Tyres Analysed"), "blue"],
            [kpis.critical, __("Critical"), "red"],
            [kpis.warning, __("Warning"), "orange"],
            [kpis.due_within_30_days, __("Due Within 30 Days"), "red"],
            [kpis.below_25_percent, __("Below 25% Tread"), "orange"],
            [kpis.underinflated, __("Underinflated >10%"), "orange"],
        ];
        $("#tyre-kpis").html(cards.map(([value, label, colour]) => `
            <div class="tyre-kpi ${colour}">
                <div class="value">${escape(value)}</div>
                <div class="label">${escape(label)}</div>
            </div>
        `).join(""));
    }

    function chart(selector, items, type, colour, datasetName) {
        $(selector).empty();
        new frappe.Chart(selector, {
            data: {
                labels: items.map(item => item.label),
                datasets: [{ name: datasetName, values: items.map(item => item.value) }],
            },
            type,
            height: 250,
            colors: colour,
            axisOptions: { xIsSeries: true },
            barOptions: { spaceRatio: 0.35 },
        });
    }

    function renderMockGraderDamage(items) {
        const selector = "#mock-grader-damage-chart";
        $(selector).empty();

        if (!items.length) {
            $(selector).html(`<div class="dashboard-empty">${__("Mock comparison data has not been installed.")}</div>`);
            return;
        }

        new frappe.Chart(selector, {
            data: {
                labels: items.map(item => item.label),
                datasets: [
                    {
                        name: __("Mock damaged tyres"),
                        chartType: "bar",
                        values: items.map(item => item.damaged_tyres),
                    },
                    {
                        name: __("Mock grader utilisation %"),
                        chartType: "line",
                        values: items.map(item => item.grader_utilisation),
                    },
                ],
            },
            type: "axis-mixed",
            height: 280,
            colors: ["#d32f2f", "#1565c0"],
            axisOptions: { xIsSeries: true },
            barOptions: { spaceRatio: 0.4 },
            lineOptions: { regionFill: 0, hideDots: 0 },
        });
    }

    function tyreCard(tyre) {
        const brand = [tyre.tyre_make, tyre.tread_pattern].filter(Boolean).join(" · ") || "—";
        const remainingLife = tyre.remaining_months === null || tyre.remaining_months === undefined
            ? "—"
            : `${tyre.remaining_months} months`;
        const pressure = tyre.actual_pressure || tyre.recommended_pressure
            ? `${tyre.actual_pressure ?? "—"} / ${tyre.recommended_pressure ?? "—"}`
            : "—";

        return `
            <div class="adt-tyre-card ${bandClass(tyre.urgency_band)}">
                <div class="title">${escape(tyre.position)} — ${escape(tyre.urgency_band)}</div>
                <ul class="spec-list">
                    <li><strong>${__("Serial")}:</strong> ${escape(tyre.serial_number || "—")}</li>
                    <li><strong>${__("Brand & pattern")}:</strong> ${escape(brand)}</li>
                    <li><strong>${__("OTD")}:</strong> ${escape(tyre.otd ?? "—")} mm</li>
                    <li><strong>${__("RTD")}:</strong> ${escape(tyre.average_rtd ?? "—")} mm (${escape(tyre.rtd_percent ?? "—")}%)</li>
                    <li><strong>${__("Wear rate")}:</strong> ${escape(tyre.wear_rate ?? "—")} mm/month</li>
                    <li><strong>${__("Pressure actual / recommended")}:</strong> ${escape(pressure)}</li>
                    <li><strong>${__("Scrap limit")}:</strong> ${escape(tyre.scrap_limit_mm)} mm</li>
                    <li><strong>${__("Replace by")}:</strong> ${escape(displayDate(tyre.replacement_date))}</li>
                    <li><strong>${__("Remaining life")}:</strong> ${escape(remainingLife)}</li>
                </ul>
            </div>
        `;
    }

    function renderAdt(view) {
        if (!view.fleet_number) {
            $("#adt-layout").html(`<div class="dashboard-empty">${__("Select an ADT above to display its tyre layout and latest readings.")}</div>`);
            return;
        }

        const special = view.tyre_layout === "b60e" || view.tyres.some(tyre => ["RRO", "RRI", "LRI", "LRO"].includes(String(tyre.position).toUpperCase()));
        const leftOrder = special ? ["LF", "LRO", "LRI"] : ["LF", "LM", "LR"];
        const rightOrder = special ? ["RF", "RRI", "RRO"] : ["RF", "RM", "RR"];
        const order = positions => positions.map(position => view.tyres.find(tyre => String(tyre.position).toUpperCase() === position)).filter(Boolean);
        const left = order(leftOrder);
        const right = order(rightOrder);
        const assigned = new Set([...left, ...right].map(tyre => tyre.serial_number));
        const extras = view.tyres.filter(tyre => !assigned.has(tyre.serial_number));
        extras.forEach((tyre, index) => (index % 2 ? right : left).push(tyre));

        $("#adt-layout").html(`
            <div class="adt-layout">
                <div class="adt-card-column">${left.map(tyreCard).join("")}</div>
                <div>
                    <div style="text-align:center;font-weight:700;margin-bottom:8px;">
                        ${escape(view.fleet_number)} · ${escape(view.vehicle_make)} ${escape(view.model)} · ${escape(view.site)}
                    </div>
                    <img class="adt-layout-image" src="${escape(view.layout_image)}" alt="ADT tyre layout">
                    ${view.tyres.length ? "" : `<div class="dashboard-empty">${__("No latest tyre survey was found for this ADT.")}</div>`}
                </div>
                <div class="adt-card-column">${right.map(tyreCard).join("")}</div>
            </div>
        `);
    }

    function renderUrgent(rows) {
        if (!rows.length) {
            $("#urgent-list").html(`<div class="dashboard-empty">${__("No tyre records were found.")}</div>`);
            return;
        }

        $("#urgent-list").html(`
            <div style="overflow:auto;">
                <table class="urgent-table">
                    <thead><tr>
                        <th>#</th><th>${__("ADT")}</th><th>${__("Site")}</th><th>${__("Position")}</th>
                        <th>${__("Serial")}</th><th>${__("RTD")}</th><th>${__("Wear")}</th>
                        <th>${__("Months Left")}</th><th>${__("Replace By")}</th><th>${__("Band")}</th>
                    </tr></thead>
                    <tbody>${rows.map((row, index) => `
                        <tr>
                            <td>${index + 1}</td><td>${escape(row.fleet_number)}</td><td>${escape(row.site)}</td>
                            <td>${escape(row.position)}</td><td>${escape(row.serial_number)}</td>
                            <td>${escape(row.average_rtd)} mm</td><td>${escape(row.wear_rate)}</td>
                            <td>${escape(row.remaining_months ?? "—")}</td><td>${escape(displayDate(row.replacement_date))}</td>
                            <td class="band ${bandClass(row.urgency_band)}">${escape(row.urgency_band)}</td>
                        </tr>
                    `).join("")}</tbody>
                </table>
            </div>
        `);
    }

    function render(data) {
        renderKpis(data.kpis);
        renderAdt(data.adt_view);
        chart("#forecast-chart", data.replacement_forecast, "bar", ["#e0a800"], __("Tyres"));
        chart("#urgency-chart", data.urgency_bands, "bar", ["#d32f2f"], __("Tyres"));
        chart("#site-wear-chart", data.site_wear, "bar", ["#1565c0"], __("mm/month"));
        chart("#brand-wear-chart", data.brand_wear, "bar", ["#00897b"], __("mm/month"));
        chart("#pressure-chart", data.pressure_compliance, "bar", ["#2e7d32"], __("Compliance %"));
        renderMockGraderDamage(data.mock_grader_damage || []);
        renderUrgent(data.urgent_tyres);
    }

    function loadDashboard() {
        frappe.call({
            method: "engineering.engineering.page.tyre_dashboard.tyre_dashboard.get_dashboard_data",
            args: {
                as_on_date: asOnDate.get_value(),
                site: site.get_value(),
                fleet_number: fleetNumber.get_value(),
            },
            freeze: true,
            freeze_message: __("Loading tyre dashboard..."),
        }).then(response => render(response.message));
    }

    loadDashboard();
};
